### PR TITLE
Web auth 1

### PR DESCRIPTION
**Pull request by murbar**

### CODE REVIEW

**Comment by murbar** [comment]
Didn't have time to finish the client app. Not really sure how to get cookies and all that with axios. I could look it up but I'm spent for today. I worked with JWT during build week and feel comfortable with them.

File: client/src/components/ErrorMessage.jsx [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+  background: tomato;
+  color: black;
+  padding: 0.5rem;
+  margin: 1rem 0;
+  border-radius: 0.5rem;
+`;
+
+const ErrorMessage = ({ children }) => {
+  return <ErrorWrapper>{children}</ErrorWrapper>;
+};
```
Comment (DaftBeowulf) [inline client/src/components/ErrorMessage.jsx]: cleanly-done error component :+1:

File: data/validators.js [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+    .label('First name')
+    .required(),
+  last_name: Joi.string()
+    .label('Last name')
+    .required()
+};
+
+module.exports = {
+  user: user => Joi.validate(user, userSchema)
+};
```
Comment (DaftBeowulf) [inline data/validators.js]: a relatively long validation function based on all the requirements you set up for registration/logging in -- the perfect example of why you would abstract it out into a separate file for later use like this :+1:

File: middleware/restricted-route.js [hunk 3]
```diff
@@ -0,0 +1,11 @@
+module.exports = (req, res, next) => {
+  try {
+    if (req && req.session && req.session.user) {
```
Comment (DaftBeowulf) [inline middleware/restricted-route.js]: in the spirit of Luis' guidance, being _extra_  safe with the triple validation here, I like it

File: routes/auth.js [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+    }
+  } catch (error) {
+    // console.log(error);
+    res.status(500).json(error);
+  }
+});
+
+router.get('/logout', async (req, res) => {
+  try {
+    req.session && (await req.session.destroy());
```
Comment (DaftBeowulf) [inline routes/auth.js]: really like the logical short-circuiting here

**Pull request merged**